### PR TITLE
[fern] Coord-norm fix replay on 5L/256d + T_max=50 (Wave 10)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Your PR #360 (`q40rez85`) validated the FourierEmbed coordinate normalization fix — ep5=9.03% (PASS) vs 9.5% gate, with consistent improvement over the no-fix baseline. The fix is real and confirmed.

Now we need to replay this fix on the **current best architecture** (5L/256d + T_max=50) rather than the 4L/256d baseline you used for verification. The coord-norm fix was validated on 4L/256d. The hypothesis is that it will provide a ~0.3pp additive gain on top of the 5L/256d+T_max=50 recipe, potentially pushing us below 6.9% abupt.

Importantly: the coord-norm fix is implemented in `model.py` (the FourierEmbed class). The flag `--fourier-pe` enables it. Make sure your code has the coord-norm fix committed and that `n_params` matches the 5L/256d expectation.

**This is a verification + improvement run, not a new invention.** Run cleanly, report clearly.

## Instructions

Confirm the coord-norm fix is still in the codebase from your PR #360 work. Then run a single 50-epoch DDP4 job:

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --wandb-group bengio-wave10-coord-norm
```

In your first PR comment, confirm:
1. The coord-norm fix is active (copy-paste the relevant FourierEmbed code)
2. The W&B run ID and n_params

**Kill thresholds** (kill if metric NOT below value at epoch):
- ep5: abupt < 12.0%
- ep15: abupt < 9.0%
- ep30: abupt < 7.2091% (must beat official baseline)

**COMPLIANCE REQUIREMENT**: Run ONLY the single command above. Do NOT launch additional arms or variants. Note any ideas in PR comments — do NOT start a second run. Unauthorized experiments will result in immediate PR closure per established lab policy.

Report per-epoch metrics at ep5, ep15, ep30, and final.

## Baseline

Current best (official BASELINE.md, PR #74):
- `val_primary/abupt_axis_mean_rel_l2_pct`: **7.2091%** (ep30, run `m9775k1v`)
- `val_primary/surface_pressure_rel_l2_pct`: 4.802%
- `val_primary/volume_pressure_rel_l2_pct`: 4.166%
- `val_primary/wall_shear_y_rel_l2_pct`: 9.100%
- `val_primary/wall_shear_z_rel_l2_pct`: 10.869%

In-flight best (alphonse #174 `vu4jsiic`, ep~42, still running):
- abupt: **6.9801%**

Your PR #360 reference (4L/256d + coord-norm fix, run `q40rez85`):
- ep5: abupt=9.03% (vs 4L/256d no-fix baseline ~9.5% at ep5)

This run targets 5L/256d + T_max=50 + coord-norm fix: expected ~6.7–6.9% at ep50.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Clean hypothesis — you validated the coord-norm fix on 4L/256d in PR #360 (ep5=9.03% vs ~9.5% baseline), and now we want to see if it compounds with the 5L/256d + T_max=50 recipe. The expected range of 6.7–6.9% at ep50 is plausible.

**Before starting:**
1. Confirm the coord-norm fix is still in your `model.py` FourierEmbed class from PR #360. Copy-paste the relevant FourierEmbed code in your first compliance comment so we can verify it is the same fix.
2. Confirm you are on the `bengio` branch with `--fourier-pe`, `--lr-cosine-t-max`, and `--no-use-ema` flags available.

**Compliance reply required:** Post your W&B run ID, `n_params`, and the FourierEmbed code snippet within 30 minutes of starting the run.

**Key monitoring checkpoints:**
- ep5: abupt < 12.0%  
- ep15: abupt < 9.0%  
- ep30: abupt < 7.2091% (must beat official baseline)  

Report metrics at each gate in PR comments. Do NOT wait until ep50 to post results.

**One run only.** Note any observations in comments — do not launch variants.

**What to report when done:**
1. W&B run ID  
2. Best val_abupt epoch and value  
3. Full test metrics from best-val checkpoint (all axes)  
4. Comparison to PR #360 4L/256d result: did the coord-norm fix provide similar relative improvement on 5L/256d?  
5. Were val slopes still negative at ep50?

**Merge bar:** val_abupt < 7.2091% (official BASELINE.md, PR #74).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #398 because ADVISOR-ENFORCEMENT (Wave 10 ACK deadline expired)

Assigned 14:05Z with 30-min ACK deadline. Now 14:49Z (~45min past deadline). Zero student comments, zero ACK.

PR closed. Coord-norm fix replay on 5L/256d + T_max=50 hypothesis remains valid for reassignment..

**Comment by morganmcg1** [comment]
STUDENT fern:

## Compliance: Coord-Norm Fix Applied + Run Launched

Cherry-picked PR #360's coord-norm fix (commit `da4975b`) onto this branch — the fix was **not** present on `bengio` (PR #360 was closed, not merged). Applied as `6a8dde3` on this branch.

### W&B run ID and n_params
- **Run ID**: `bar2r8ha` (rank0)
- **All ranks**: rank0=`bar2r8ha`, rank1=`oegeph6b`, rank2=`4ddojkle`, rank3=`x8st3d9s`
- **W&B URL**: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/bar2r8ha
- **Group**: `bengio-wave10-coord-norm`
- **n_params**: **3.99M** (3,992,313) — matches 5L/256d expectation (4L/256d baseline `m9775k1v` was 3.249M; +0.74M aligns with one extra Transolver block ≈ 0.73M).

### FourierEmbed code (active)
```python
class FourierEmbed(nn.Module):
    """NeRF-style multi-scale sinusoidal positional encoding for 3D coordinates.

    Coordinates are normalized to roughly [-1, 1] with a fixed DrivAerML
    domain bounding box (5-sigma envelope on volume_xyz, padded slightly),
    then encoded with a base-2 geometric frequency progression
    ``pi * 2^k``, ``k in [0, num_freqs)``. Sin/cos features are concatenated
    and projected to hidden_dim with a learned linear layer.
    """

    DOMAIN_BOUNDS = (
        (-12.0, 14.0),
        (-5.0, 5.0),
        (-3.0, 3.0),
    )

    def __init__(self, hidden_dim: int, input_dim: int = 3, num_freqs: int = 8):
        super().__init__()
        if not 1 <= input_dim <= len(self.DOMAIN_BOUNDS):
            raise ValueError(
                f"input_dim must be in [1, {len(self.DOMAIN_BOUNDS)}]; got {input_dim}"
            )
        if num_freqs <= 0:
            raise ValueError("num_freqs must be positive")
        self.hidden_dim = hidden_dim
        self.input_dim = input_dim
        self.num_freqs = num_freqs
        bounds = self.DOMAIN_BOUNDS[:input_dim]
        center = torch.tensor([(lo + hi) / 2.0 for lo, hi in bounds], dtype=torch.float32)
        half_range = torch.tensor([(hi - lo) / 2.0 for lo, hi in bounds], dtype=torch.float32)
        self.register_buffer("center", center)
        self.register_buffer("half_range", half_range)
        freqs = math.pi * (2.0 ** torch.arange(num_freqs, dtype=torch.float32))
        self.register_buffer("freqs", freqs)
        raw_dim = input_dim * num_freqs * 2
        self.proj = LinearProjection(raw_dim, hidden_dim)

    def forward(self, coords: torch.Tensor) -> torch.Tensor:
        coords = coords.float()
        coords = (coords - self.center) / self.half_range
        angles = coords.unsqueeze(-1) * self.freqs
        emb = torch.cat([torch.sin(angles), torch.cos(angles)], dim=-1)
        emb = emb.flatten(start_dim=-2)
        return self.proj(emb)
```

### Heads-up — fix was not on bengio
The advisor PR body says "Confirm the coord-norm fix is still in the codebase from your PR #360 work" — it wasn't; `da4975b` was only on the closed `fern/fouriembed-coord-normalization-fix` branch and never merged. I cherry-picked it onto this branch as a separate, isolated commit so the diff is exactly the fix and nothing else. Worth checking whether other in-flight Wave 9/10 PRs (alphonse, gilbert) also need this same cherry-pick — I can confirm from the `model.py` HEAD on `bengio` that they don't have it.

### Run command
```bash
torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema \
  --lr 3e-4 --lr-cosine-t-max 50 --no-compile-model \
  --agent fern --wandb-name "fern/coord-norm-5l256d-tmax50" \
  --wandb-group bengio-wave10-coord-norm
```

Will report at ep5, ep15, ep30 gates.
